pkg/generators/openapi_test.go

```diff
@@ -1785,3 +1785,118 @@ type Blah struct {
 	}
 
 }
+
+// Show that types with unmarshalJSON in their hierarchy do not have struct
+// defaults enforced, and that aliases and embededd types are respected
+func TestMustEnforceDefaultStruct(t *testing.T) {
+	callErr, funcErr, assert, _, funcBuffer, imports := testOpenAPITypeWriter(t, `
+package foo
+
+type Time struct {
+	value interface{}
+}
+
+
+type TimeWithoutUnmarshal struct {
+	value interface{}
+}
+
+func (_ TimeWithoutUnmarshal) OpenAPISchemaType() []string { return []string{"string"} }
+func (_ TimeWithoutUnmarshal) OpenAPISchemaFormat() string { return "date-time" }
+
+func (_ Time) UnmarshalJSON([]byte) error {
+	return nil
+}
+
+
+func (_ Time) OpenAPISchemaType() []string { return []string{"string"} }
+func (_ Time) OpenAPISchemaFormat() string { return "date-time" }
+
+// Time with UnmarshalJSON defined on pointer instead of struct
+type MicroTime struct {
+	value interface{}
+}
+
+func (t *MicroTime) UnmarshalJSON([]byte) error {
+	return nil
+}
+
+func (_ MicroTime) OpenAPISchemaType() []string { return []string{"string"} }
+func (_ MicroTime) OpenAPISchemaFormat() string { return "date-time" }
+
+type Int64 int64
+
+type Duration struct {
+	Int64
+}
+
+func (_ Duration) OpenAPISchemaType() []string { return []string{"string"} }
+func (_ Duration) OpenAPISchemaFormat() string { return "" }
+
+type NothingSpecial struct {
+	Field string
+}
+
+// +k8s:openapi-gen=true
+type Blah struct {
+	Embedded Duration
+	PointerUnmarshal MicroTime
+	StructUnmarshal Time
+	NoUnmarshal TimeWithoutUnmarshal
+	Regular NothingSpecial
+}
+	`)
+	assert.NoError(funcErr)
+	assert.NoError(callErr)
+	assert.ElementsMatch(imports, []string{`foo "base/foo"`, `common "k8s.io/kube-openapi/pkg/common"`, `spec "k8s.io/kube-openapi/pkg/validation/spec"`})
+
+	if formatted, err := format.Source(funcBuffer.Bytes()); err != nil {
+		t.Fatal(err)
+	} else {
+		assert.Equal(string(formatted), `func schema_base_foo_Blah(ref common.ReferenceCallback) common.OpenAPIDefinition {
+	return common.OpenAPIDefinition{
+		Schema: spec.Schema{
+			SchemaProps: spec.SchemaProps{
+				Type: []string{"object"},
+				Properties: map[string]spec.Schema{
+					"Embedded": {
+						SchemaProps: spec.SchemaProps{
+							Default: 0,
+							Ref:     ref("base/foo.Duration"),
+						},
+					},
+					"PointerUnmarshal": {
+						SchemaProps: spec.SchemaProps{
+							Ref: ref("base/foo.MicroTime"),
+						},
+					},
+					"StructUnmarshal": {
+						SchemaProps: spec.SchemaProps{
+							Ref: ref("base/foo.Time"),
+						},
+					},
+					"NoUnmarshal": {
+						SchemaProps: spec.SchemaProps{
+							Default: map[string]interface{}{},
```

Comment:
@apelisse I don't think there are current upstream examples of this but if you override `OpenAPISchemaType` without unmarshal JSON you still get the struct default.

I'm thinking maybe should switch current code to check for OpenAPISchemaType instead of UnmarshalJSON, since we are trying to prevent inconsistent `type` + `default`. WDYT?

Reply:
Also curious why `OpenAPISchemaType` isnt just a marker comment

Reply:
Because we haven't had the need to do it, and we probably don't want people to annotate the type when we can guess it from the actual type?